### PR TITLE
Configure to test webhooks locally

### PR DESCRIPTION
**Pull request by rabi**
This target will generate self-signed certs and configure the webhook for it to be tested locally with 'make run-with-webhook'.

### CODE REVIEW

File: hack/configure_local_webhook.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+set -ex
+
+TMPDIR=${TMPDIR:-"/tmp/k8s-webhook-server/serving-certs"}
+
+mkdir -p ${TMPDIR}
+rm -rf ${TMPDIR}/* || true
+
+openssl req -newkey rsa:2048 -days 3650 -nodes -x509 \
```
Comment (bogdando) [inline hack/configure_local_webhook.sh:9]: would it make sense to use cert-manager for development purposes only, and just issue ACME certs with letsencrypt?

I know its inclusion under the product is under question, just I believe we still could use it for testing purposes, like this. And avoid all the manual certs management

**Review by bogdando — APPROVED** [review]

**Comment by abays** [comment]
I tried this and it doesn't seem to work in my environment.  Perhaps I am doing something wrong.  Here are my steps:

1. Deploy OpenStack operator via `make openstack`
2. Remove `.spec.webhookdefinitions` and OpenStack operator controller-manager from CSV
3. (optionally delete existing OpenStack operator controller-manager deployment because sometimes removing it from the CSV doesn't cause the existing pod to terminate)
4. Execute `make run-with-webhooks`
5. Attempt to apply `OpenStackControlPlane` via `make openstack_deploy`

I see this when I try step 5:

```
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "vopenstackcontrolplane.kb.io": failed to call webhook: Post "https://ajbshuttle:9443/validate-core-openstack-org-v1beta1-openstackcontrolplane?timeout=10s": dial tcp 192.168.1.212:9443: connect: connection refused
```

**Comment by rabi** [comment]
> I tried this and it doesn't seem to work in my environment. 

I guess you're using openshift/OKD cluster and not crc?  Is ajbshuttle your $HOSTNAME? Looks like it's resolving to 192.168.1.212 but that's not the externally (from the cluster) reachable ip of your host where you're running 'make run-with-webhook'. Do you've some proxy already running or something(?).

**Comment by abays** [comment]
> > I tried this and it doesn't seem to work in my environment.
> 
> Are you using openshift/OKD cluster or crc? Is ajbshuttle your $HOSTNAME? Looks like it's resolving to 192.168.1.212 but that's not the externally (from the cluster) reachable ip of your host where you're running 'make run-with-webhook'. Do you've some proxy/sshuttle already running or something(?).
> 
> I run `make run-with-webhook` from the host running crc vm and it works fine. For other cases the local controller host (in your case probably the local machine using sshuttle) should be reachabable from the api server/cluster.

`ajbshuttle` is my workstation where I am running CRC.  I also tried on my dedicated server that runs a full (virtualized) OCP cluster, and I didn't have any luck there either.  There are no proxies running, so I'll just tinker with my environment to figure out why it behaves like it does.

**Comment by rabi** [comment]
Maybe firewalld is blocking the port on your workstation. Below may help.

[stack@osp-df-2 ~]$ sudo firewall-cmd --list-ports
80/tcp 443/tcp 6443/tcp
[stack@osp-df-2 ~]$ sudo firewall-cmd --add-port=9443/tcp
success
[stack@osp-df-2 ~]$ sudo firewall-cmd --runtime-to-permanent
success

**Comment by dprince** [comment]
@rabi this doesn't work for me either. My use case would be:

make openstack
oc edit csv (set replicase for openstack-controller-manager to 0)
make run-with-webhooks

After doing the following you won't get access to the webhooks API:

$ oc create -f config/samples/core_v1beta1_openstackcontrolplane.yaml
Error from server (InternalError): error when creating "config/samples/core_v1beta1_openstackcontrolplane.yaml": Internal error occurred: failed calling webhook "vopenstackcontrolplane.kb.io": failed to call webhook: Post "https://openstack-operator-controller-manager-service.openstack.svc:443/validate-core-openstack-org-v1beta1-openstackcontrolplane?timeout=10s": no endpoints available for service "openstack-operator-controller-manager-service"

I think we are missing some proxying here to make this work.

**Comment by dprince** [comment]
@rabi and after removing the webhooks with

oc patch csv openstack-operator.v0.0.1 --type=json -p="[{'op': 'remove', 'path': '/spec/webhookdefinitions'}]"

Still hitting:
[dprince@dovetail openstack-operator]$ oc create -f config/samples/core_v1beta1_openstackcontrolplane.yaml
Error from server (InternalError): error when creating "config/samples/core_v1beta1_openstackcontrolplane.yaml": Internal error occurred: failed calling webhook "vopenstackcontrolplane.kb.io": failed to call webhook: Post "https://dovetail.local:9443/validate-core-openstack-org-v1beta1-openstackcontrolplane?timeout=10s": dial tcp 192.168.2.119:9443: connect: connection refused

**Review by abays — COMMENTED** [review]
+1 from me.  The latest iteration works in my environments.

**Pull request merged**